locator/topology.cc

```diff
@@ -25,81 +25,264 @@ thread_local const endpoint_dc_rack endpoint_dc_rack::default_location = {
+        ret._dc_rack_nodes = _dc_rack_nodes;
+        ret._dc_endpoints = _dc_endpoints;
+        ret._dc_racks = _dc_racks;
+        ret._datacenters = _datacenters;
+    } else {
+        tlogger.debug("topology[{}]: clone_gently to {} from shard {}", fmt::ptr(this), fmt::ptr(&ret), _shard);
+        for (const auto& n : _all_nodes) {
+            ret.add_node(make_lw_shared<node>(*n));
+            co_await coroutine::maybe_yield();
+        }
+        // local node may be detached for _all_nodes
+        // if it was decommissioned.
+        if (!ret._local_node && _local_node) {
+            ret.add_node(make_lw_shared<node>(*_local_node));
         }
     }
     co_await coroutine::maybe_yield();
-    ret._current_locations.reserve(_current_locations.size());
-    for (const auto& p : _current_locations) {
-        ret._current_locations.emplace(p);
-    }
-    co_await coroutine::maybe_yield();
-    ret._datacenters = _datacenters;
     ret._sort_by_proximity = _sort_by_proximity;
     co_return ret;
 }
 
-void topology::update_endpoint(const inet_address& ep, endpoint_dc_rack dr)
-{
-    auto current = _current_locations.find(ep);
+node_ptr topology::add_node(host_id id, const inet_address& ep, const endpoint_dc_rack& dr) {
+    if (dr.dc.empty() || dr.rack.empty()) {
+        on_internal_error(tlogger, "Node must have valid dc and rack");
+    }
+    auto is_local = node::local(ep == utils::fb_utilities::get_broadcast_address());
+    if (is_local && _local_node) {
+        if (_local_node->host_id() == id) {
+            on_internal_error_noexcept(tlogger, format("Local node already set: host_id={} endpoint={} dc={} rack={}: currently mapped to host_id={} endpoint={}",
+                    id, ep, dr.dc, dr.rack,
+                    _local_node->host_id(), _local_node->endpoint()));
+            return _local_node;
+        }
+        // Replacing node with the same ip address
+        is_local = node::local::no;
+    }
+    return add_node(make_lw_shared<node>(id, ep, dr, is_local));
+}
 
-    if (current != _current_locations.end()) {
-        if (current->second.dc == dr.dc && current->second.rack == dr.rack) {
-            return;
+node_ptr topology::add_node(mutable_node_ptr node) {
+    tlogger.debug("topology[{}]: add_node: node={} host_id={} endpoint={} dc={} rack={} local={}, at {}", fmt::ptr(this), fmt::ptr(node.get()),
+            node->host_id(), node->endpoint(), node->dc_rack().dc, node->dc_rack().rack, node->is_local(), current_backtrace());
+    if (node->is_local() && _local_node && _local_node != node) {
+        on_internal_error(tlogger, format("Local node already set: host_id={} endpoint={} dc={} rack={}: currently mapped to host_id={} endpoint={}",
+                node->host_id(), node->endpoint(), node->dc_rack().dc, node->dc_rack().rack,
+                _local_node->host_id(), _local_node->endpoint()));
+    }
+    if (_all_nodes.contains(node)) {
+        return node;
+    }
```

Comment:
Shouldn't we treat adding the local node again (above) the same?

Reply:
If it's a different local node we call on_internal_error
if it's the same local node, `_all_nodes` contains it too, so it will be treated the same way.